agent/agent.go

```diff
@@ -113,143 +110,52 @@ type Config struct {
-	if err != nil {
-		return nil, trace.Wrap(err, "failed to update serf agent tags")
-	}
-
-	metricsListener, err := net.Listen("tcp", config.MetricsAddr)
-	if err != nil {
-		return nil, trace.Wrap(err, "failed to serve prometheus metrics")
-	}
-
-	ctx, cancel := context.WithTimeout(context.Background(), timelineInitTimeout)
-	defer cancel()
-
-	sqliteConfig := sqlite.Config{
-		DBPath:            config.DBPath,
-		RetentionDuration: config.RetentionDuration,
-		Clock:             config.Clock,
-	}
-	timeline, err := sqlite.NewTimeline(ctx, sqliteConfig)
-	if err != nil {
-		return nil, trace.Wrap(err, "failed to initialize timeline")
-	}
-
-	agent := &agent{
-		Config:                  *config,
-		SerfClient:              client,
-		Timeline:                timeline,
-		name:                    config.Name,
-		cache:                   config.Cache,
-		dialRPC:                 DefaultDialRPC(config.CAFile, config.CertFile, config.KeyFile),
-		statusClock:             config.Clock,
-		recycleClock:            config.Clock,
-		timelineClock:           config.Clock,
-		statusQueryReplyTimeout: statusQueryReplyTimeout,
-		localStatus:             emptyNodeStatus(config.Name),
-		metricsListener:         metricsListener,
-		done:                    make(chan struct{}),
-	}
-
-	agent.rpc, err = newRPCServer(agent, config.CAFile, config.CertFile, config.KeyFile, config.RPCAddrs)
-	if err != nil {
-		return nil, trace.Wrap(err, "failed to create RPC server")
-	}
-	return agent, nil
-}
-
 // CheckAndSetDefaults validates this configuration object.
 // Config values that were not specified will be set to their default values if
 // available.
 func (r *Config) CheckAndSetDefaults() error {
 	var errors []error
-
 	if r.CAFile == "" {
 		errors = append(errors, trace.BadParameter("certificate authority file must be provided"))
 	}
-
 	if r.CertFile == "" {
 		errors = append(errors, trace.BadParameter("certificate must be provided"))
 	}
-
 	if r.KeyFile == "" {
 		errors = append(errors, trace.BadParameter("certificate key must be provided"))
 	}
-
 	if r.Name == "" {
 		errors = append(errors, trace.BadParameter("agent name cannot be empty"))
 	}
-
 	if len(r.RPCAddrs) == 0 {
 		errors = append(errors, trace.BadParameter("at least one RPC address must be provided"))
 	}
-
-	if r.DBPath == "" {
-		errors = append(errors, trace.BadParameter("sqlite database path must be provided"))
-	}
-
 	if r.Tags == nil {
 		r.Tags = make(map[string]string)
 	}
-
 	if r.Clock == nil {
 		r.Clock = clockwork.NewRealClock()
 	}
-
 	return trace.NewAggregate(errors...)
 }
 
 type agent struct {
+	sync.Mutex
 	health.Checkers
 
 	metricsListener net.Listener
 
-	// SerfClient provides access to the serf agent.
-	SerfClient SerfClient
-
-	// Name of this agent.  Must be the same as the serf agent's name
-	// running on the same node.
-	name string
-
 	// RPC server used by agent for client communication as well as
 	// status sync with other agents.
 	rpc RPCServer
```

Comment:
Does agent need a reference to the RPC server?
From what I can see agent only uses this reference to stop the server in the `Close` method.

Reply:
Not sure what you mean?
If you meant it could do away with a simple stopper interface - then sure, it could.

Reply:
Sorry, should have elaborate a bit more on this.

- agent stores a pointer to the RPC server.
- RPC server also stores a pointer to the agent.
- ^ Unnecessary circular reference.
- The server de-references the agent when handling RPCs, so the server needs a pointer to the agent.
- The agent only holds a pointer to the server so that it can shutdown the server.

```golang
func runAgent(config *agent.Config, monitoringConfig *config, peers []string) error {
	...
	monitoringAgent, err := agent.New(config)
	if err != nil {
		return trace.Wrap(err)
	}
	defer monitoringAgent.Close()


	if err = addCheckers(monitoringAgent, monitoringConfig); err != nil {
		return trace.Wrap(err)
	}
	if err = monitoringAgent.Start(); err != nil {
		return trace.Wrap(err)
	}
	if len(peers) > 0 {
		if err = monitoringAgent.Join(peers); err != nil {
			return trace.Wrap(err, "failed to join serf cluster")
		}
	}
        ...
}
```
Currently it looks like the main function handles agent management (init/start/stop).
But I feel like it would make more sense to have to have it the other way. Main should handle server management (init/start/stop) and have the server manage agent internally. This way, agent does not need to store a pointer to the server.

```golang
func runServer(...) error {
    monitoringAgent, err := agent.New(...)
    if err != nil {
        return trace.Wrap(err)
    }
    // Additional agent init
    ...

    server, err := newRPCServer(agent, ...)
    if err != nil {
        return trace.Wrap(err)
    }
    // Handle server start/stop
    ....
}
```